src/Controller/Component/ModulesComponent.php

```diff
@@ -145,10 +145,40 @@ function (&$data, $key) use ($metaModules) {
             array_diff_key($metaModules, $modules),
             $pluginModules
         );
+        $this->modules = $this->modulesByRoleConfig($this->modules);
 
         return $this->modules;
     }
 
+    /**
+     * This filters modules and apply 'Role' config by user role, if any.
+     * Module can be "hidden": remove from $modules.
+     * Module can be "readonly": add "'readonly' => true" to module.
+     *
+     * @param array $modules The modules
+     * @return array
+     */
+    protected function modulesByRoleConfig(array $modules): array
+    {
+        $user = $this->getController()->Auth->user();
+        $roles = (array)Hash::get($user, 'roles');
+        $hidden = [];
+        $readonly = [];
+        foreach ($roles as $role) {
+            $hidden = array_merge($hidden, (array)Configure::read(sprintf('AccessControl.%s.hidden', $role)));
```

Comment:
not sure about `array_merge` => if a module is `hidden` on a role but not in another I think it should not be hidden